src/relay/pass/well_formed.cc

```diff
@@ -0,0 +1,50 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file well_formed.cc
+ * \brief check that expression is well formed.
+ */
+#include <tvm/relay/pass.h>
+#include <tvm/relay/expr_functor.h>
+#include <unordered_set>
+
+namespace tvm {
+namespace relay {
+
+struct NotWellFormed { };
+
+//! brief make sure each Var is bind at most once.
+struct IsWellForm : ExprVisitor {
```

Comment:
we do not need an exception in here. you can directly store a flag in the case and return as soon as the flag is used. Just to minimize the use of un-necessary exceptions. c.f. https://github.com/dmlc/tvm/blob/master/src/pass/ssa.cc#L178

Reply:
yes, but exceptioning will be quicker, as we need not traverse the whole tree.